Cubical/Structures/Semigroup.agda

```diff
@@ -3,80 +3,160 @@ module Cubical.Structures.Semigroup where
 
 open import Cubical.Foundations.Prelude
 open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.Equiv.HalfAdjoint
 open import Cubical.Foundations.HLevels
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.Univalence
+open import Cubical.Foundations.Transport
+open import Cubical.Foundations.SIP renaming (SNS-PathP to SNS)
+
 open import Cubical.Data.Sigma
 
-open import Cubical.Foundations.SIP renaming (SNS-PathP to SNS)
 open import Cubical.Structures.Pointed
 open import Cubical.Structures.NAryOp
 
+open Iso
+
 private
   variable
-    ℓ ℓ' : Level
+    ℓ : Level
+
+-- Semigroups as a record, inspired by the Agda standard library:
+--
+--  https://github.com/agda/agda-stdlib/blob/master/src/Algebra/Bundles.agda#L48
+--  https://github.com/agda/agda-stdlib/blob/master/src/Algebra/Structures.agda#L50
+--
+-- Note that as we are using Path for all equations the IsMagma record
+-- would only contain isSet A if we had it.
+record IsSemigroup {A : Type ℓ} (_·_ : A → A → A) : Type ℓ where
+
+  constructor issemigroup
 
-raw-semigroup-structure : Type ℓ → Type ℓ
-raw-semigroup-structure X = X → X → X
+  field
+    is-set : isSet A
+    assoc  : (x y z : A) → x · (y · z) ≡ (x · y) · z
 
-raw-semigroup-is-SNS : SNS {ℓ} raw-semigroup-structure _
-raw-semigroup-is-SNS = binaryFunSNS pointed-iso pointed-is-SNS
 
-semigroup-axioms : (X : Type ℓ) → raw-semigroup-structure X → Type ℓ
-semigroup-axioms X _·_ = isSet X ×
-                         ((x y z : X) → (x · (y · z)) ≡ ((x · y) · z))
+record Semigroup : Type (ℓ-suc ℓ) where
 
-semigroup-structure : Type ℓ → Type ℓ
-semigroup-structure = add-to-structure (raw-semigroup-structure) semigroup-axioms
+  constructor semigroup
 
-Semigroup : Type (ℓ-suc ℓ)
-Semigroup {ℓ} = TypeWithStr ℓ semigroup-structure
+  field
+    Carrier     : Type ℓ
+    _·_         : Carrier → Carrier → Carrier
+    isSemigroup : IsSemigroup _·_
 
--- Operations for extracting components
+  infixl 7 _·_
 
+  open IsSemigroup isSemigroup public
+
+-- Extractor for the carrier type
 ⟨_⟩ : Semigroup → Type ℓ
-⟨ G , _ ⟩ = G
+⟨_⟩ = Semigroup.Carrier
+
+
+record SemigroupIso (M N : Semigroup {ℓ}) : Type ℓ where
+
+  constructor semigroupiso
+
+  -- Shorter qualified names
+  module M = Semigroup M
+  module N = Semigroup N
```

Comment:
You might want to put these definitions in a `private`.

Reply:
Good point! Fixed